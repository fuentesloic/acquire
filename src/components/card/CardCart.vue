<template>
  <article>
    <header>
      <BaseTitle class="cart">Welcome back Sarah !</BaseTitle>
      <BaseParagraph class="cart">Nice to see you again!</BaseParagraph>
      <BaseButton class="cart" :click="() => `Congrats! you own me ${totalPrice}`">
        pay ${{ totalPrice }}
      </BaseButton>
    </header>

    <section v-if="products().length">
      <BaseDivider />
      <BaseTitle class="cart">Your cart</BaseTitle>

      <ul>
        <li v-for="product in products()" :key="product.id">
          <figure>
            <BaseImage class="cart" :alt="product.name" :src="product.image" />
          </figure>

          <figcaption>
            <BaseTitle class="cart product">{{ product.name }}</BaseTitle>
            <BaseSubtitle class="cart product">{{ product.variant.name }}</BaseSubtitle>

            <div class="ctas">
              <BaseParagraph class="cta">
                <BaseButton class="cart selector" :click="() => removeProductQtt(product)">-</BaseButton>
                <span>{{ product.quantity }}</span>
                <BaseButton class="cart selector" :click="() => addProductQtt(product)">+</BaseButton>
              </BaseParagraph>

              <BaseParagraph class="cta">$ {{ product.variant.price }}</BaseParagraph>
            </div>
          </figcaption>
        </li>
      </ul>
    </section>

    <footer>
      <BaseParagraph>Powered by Acquire</BaseParagraph>
    </footer>
  </article>
</template>

<script lang="ts">
import { mapGetters } from 'vuex'
import { Component, Vue } from 'vue-property-decorator'

import BaseButton from '@/components/base/BaseButton.vue'
import BaseDivider from '@/components/base/BaseDivider.vue'
import BaseImage from '@/components/base/BaseImage.vue'
import BaseParagraph from '@/components/base/BaseParagraph.vue'
import BaseTitle from '@/components/base/BaseTitle.vue'
import BaseSubtitle from '@/components/base/BaseSubtitle.vue'
import { ProductCart } from '@/model/cart'
import { IMutationCart } from '@/model/store'

@Component({
  components: {
    BaseButton,
    BaseDivider,
    BaseImage,
    BaseParagraph,
    BaseTitle,
    BaseSubtitle
  },
  computed: {
    ...mapGetters('cart', { totalPrice: 'totalPrice' })
  }
})
export default class CardProduct extends Vue {
  /// Computed: provide sync product list
  products(): ProductCart[] {
    return this.$store.state.cart.products
  }

  /// Method: Add product quantity
  public addProductQtt(product: ProductCart): void {
    const payload: IMutationCart['addProductQtt'] = { product }

    this.$store.commit('cart/addProductQtt', payload)
  }

  /// Method: Remove product quantity, will remove it IFF qtt === 0
  public removeProductQtt(product: ProductCart): void {
    const payload: IMutationCart['removeProductQtt'] = { product }

    this.$store.commit('cart/removeProductQtt', payload)
  }
}
</script>

<style lang="scss" scoped>
article {
  display: flex;
  flex-direction: column;
  text-align: center;
  background-color: $white;
  border-radius: 0.8rem;
  margin-bottom: 2rem;
  padding: 2rem;
  position: relative;

  section {
    text-align: left;

    li {
      display: flex;

      + li {
        margin-top: 1rem;
      }

      figure {
        background-color: $gray-light;
      }

      figcaption {
        display: flex;
        flex: 1;
        flex-direction: column;
        margin-left: 1rem;

        .ctas {
          align-items: center;
          display: flex;
          justify-content: space-between;

          span {
            margin: 0 1rem;
          }

          .cta {
            align-items: center;
            display: flex;
          }
        }
      }
    }
  }

  footer {
    bottom: -4rem;
    text-align: center;
    left: 0;
    position: absolute;
    right: 0;
  }
}
</style>
