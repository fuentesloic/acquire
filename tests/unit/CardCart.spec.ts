import Vuex from 'vuex'
import { createLocalVue, shallowMount } from '@vue/test-utils'

import CardCart from '@/components/card/CardCart.vue'
import { fakeProductCartOne, fakeProductCartTwo } from './mockup'

const localVue = createLocalVue()
localVue.use(Vuex)

function createLocalWrapper() {
  const computed = {
    products: () => [fakeProductCartOne, fakeProductCartTwo]
  }

  const methods = {
    addProductQtt: jest.fn(),
    removeProductQtt: jest.fn()
  }

  const store = new Vuex.Store({
    state: {
      products: [fakeProductCartOne, fakeProductCartTwo]
    },
    getters: {
      totalPrice: () => 42
    }
  })

  return shallowMount(CardCart, {
    store,
    localVue,
    computed,
    methods,
    propsData: {
      products: []
    }
  })
}

describe('Components - Card - CardCart', () => {
  const wrapper = createLocalWrapper()

  wrapper.setData({ products: [fakeProductCartOne, fakeProductCartTwo] })

  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders cart header', () => {
    expect(wrapper.find('article header basetitle-stub').classes('cart')).toBe(true)
    expect(wrapper.find('article header baseparagraph-stub').classes('cart')).toBe(true)
    expect(wrapper.find('article header basebutton-stub').classes('cart')).toBe(true)

    wrapper.setData({ products: [fakeProductCartOne, fakeProductCartTwo] })
    expect(wrapper.vm.$store.getters.totalPrice()).toBe(42)
  })

  it('renders cart content IFF products', () => {
    wrapper.setData({ products: [] })
    expect(wrapper.find('section header baseparagraph-stub')).toBe(true)

    wrapper.setData({ products: [fakeProductCartOne, fakeProductCartTwo] })
    expect(wrapper.find('section header baseparagraph-stub')).toBe(false)
  })

  it('renders cart footer', () => {
    expect(wrapper.find('article footer baseparagraph-stub')).toBe(true)
  })
})
