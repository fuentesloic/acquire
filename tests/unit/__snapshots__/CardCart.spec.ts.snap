// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Card - CardCart renders without crashing 1`] = `
<article products="">
  <header>
    <basetitle-stub product="" cart="" class="cart">Welcome back Sarah !</basetitle-stub>
    <baseparagraph-stub cta="" label="" class="cart">Nice to see you again!</baseparagraph-stub>
    <basebutton-stub active="" cart="" click="[Function]" selector="" class="cart">
      pay $
    </basebutton-stub>
  </header>
  <section>
    <basedivider-stub></basedivider-stub>
    <basetitle-stub product="" cart="" class="cart">Your cart</basetitle-stub>
    <ul>
      <li>
        <figure>
          <baseimage-stub alt="product_name_one" cart="" logo="" product="" src="product_image_one" class="cart"></baseimage-stub>
        </figure>
        <figcaption>
          <basetitle-stub product="" cart="" class="cart product">product_name_one</basetitle-stub>
          <basesubtitle-stub product="" cart="" class="cart product">variant_name_one</basesubtitle-stub>
          <div class="ctas">
            <baseparagraph-stub cta="" label="" class="cta">
              <basebutton-stub active="" cart="" click="[Function]" selector="" class="cart selector">-</basebutton-stub> <span>4</span>
              <basebutton-stub active="" cart="" click="[Function]" selector="" class="cart selector">+</basebutton-stub>
            </baseparagraph-stub>
            <baseparagraph-stub cta="" label="" class="cta">$ 42</baseparagraph-stub>
          </div>
        </figcaption>
      </li>
      <li>
        <figure>
          <baseimage-stub alt="product_name_two" cart="" logo="" product="" src="product_image_two" class="cart"></baseimage-stub>
        </figure>
        <figcaption>
          <basetitle-stub product="" cart="" class="cart product">product_name_two</basetitle-stub>
          <basesubtitle-stub product="" cart="" class="cart product">variant_name_two</basesubtitle-stub>
          <div class="ctas">
            <baseparagraph-stub cta="" label="" class="cta">
              <basebutton-stub active="" cart="" click="[Function]" selector="" class="cart selector">-</basebutton-stub> <span>2</span>
              <basebutton-stub active="" cart="" click="[Function]" selector="" class="cart selector">+</basebutton-stub>
            </baseparagraph-stub>
            <baseparagraph-stub cta="" label="" class="cta">$ 42</baseparagraph-stub>
          </div>
        </figcaption>
      </li>
    </ul>
  </section>
  <footer>
    <baseparagraph-stub cta="" label="">Powered by Acquire</baseparagraph-stub>
  </footer>
</article>
`;
